### ISSUE TITLE
Fine-tuning pipeline ready — branch vybn/fine-tuning-pipeline

### ISSUE THREAD

**Issue by zoedolan** [zoedolan/Vybn#2481]
## What's on the branch

`spark/fine_tuning/` contains the complete QLoRA pipeline:

1. **`prepare_training_data.py`** — Converts Vybn corpus (journals, autobiography vols I-V, reflections, memoirs, mind docs) into 594 chat-format JSONL training examples (~4.7M chars). PII-scanned, clean.

2. **`train_qlora.py`** — QLoRA training targeting attention layers only (`q_proj`, `k_proj`, `v_proj`, `o_proj`). Rank 8, alpha 16, 3 epochs, gradient checkpointing. Requires stopping vLLM first (needs the GPU memory).

3. **`merge_and_quantize.py`** — Merges LoRA adapter into the unquantized base (`MiniMaxAI/MiniMax-M2.5`, cached on host), then re-quantizes to compressed-tensors 4-bit. The BF16 merge step needs ~458GB — will use CPU/disk offload.

4. **`vybn_training_data.jsonl`** — The actual training data (594 examples).

## What I validated

- PEFT installs clean in the vLLM container
- Model config probe works: 62 layers, 256 experts, hidden_size 3072
- Attention layer names confirmed: `self_attn.{q,k,v,o}_proj.weight_packed`
- Quantization format is `compressed-tensors` (not standard AWQ) — accounted for in merge script
- Training data PII scan clean (only matches were public courthouse addresses)

## What happened with --enable-lora

Crashed with `AssertionError` in `fused_moe.py` line 161 — LoRA injection expects TritonExperts but Blackwell uses CUTLASS fused MoE. Model restored to original config, serving normally. Full analysis in #2480.

## To run the training

```bash
# 1. Stop vLLM (frees GPU memory for training)
docker exec vllm_node pkill -f 'vllm serve'

# 2. Train (inside container, where torch lives)
docker exec -it vllm_node python3 /workspace/train_qlora.py
# Or copy script into container first

# 3. Restart vLLM with merged model
# (after merge_and_quantize.py completes)
```

## Open questions for you

1. **Timing**: Training will take the model offline. Want to schedule this for a low-traffic window?
2. **Merge step memory**: The BF16 merge needs ~458GB. CPU/disk offload should work but will be slow (~hours). Alternative: cloud burst for just that step. Your call.
3. **The unquantized base model** (`MiniMaxAI/MiniMax-M2.5`) is cached at `~/.cache/huggingface/hub/` on the host. Need to verify it's complete — it's ~458GB in BF16.

Ready to merge the branch and run whenever you give the word.

**Comment by zoedolan** [comment]
Branch merged. Proceed as follows: first, verify the unquantized base model cache is complete at `~/.cache/huggingface/hub/` on the host (~458GB BF16) — run a quick size/integrity check before committing GPU time. Then stop vLLM to free GPU memory and run `train_qlora.py` inside the container (`docker exec -it vllm_node python3 /workspace/train_qlora.py`). Once training completes, assess whether the BF16 merge step (~458GB) is feasible via CPU/disk offload on current hardware or whether a short cloud burst is the cleaner path — choose whichever keeps the timeline tighter. Run `merge_and_quantize.py` accordingly, then restart vLLM pointing at the merged model. The training data is the corpus — the journals, autobiography, reflections — so this step is the one that actually matters. Take your time, do it right.
